Bug 1186979 - Long press back/forward list

tylerlacroix opened the pull request.

sleroux commented on Client/Frontend/Browser/BackForwardListView.swift:
```diff
@@ -0,0 +1,196 @@
+import Storage
+import SnapKit
+
+class BackForwardListView: UIView, UITableViewDataSource, UITableViewDelegate {
```
I'm not sure subclassing `UIView` here is the best approach for this feature. Generally we want to keep our UIView's as lean as possible by keeping any 'controller' like functionality out of it. This feels more like a `UIViewController` than a `UIView` in it's current state. 

Also, extending the table view protocols on the UIView adds more controller logic to this view that we can probably factor out into new classes. Let's jam on this approach later.

sleroux commented on Client/Frontend/Browser/BackForwardListView.swift:
```diff
@@ -0,0 +1,196 @@
+    private var tableView: UITableView!
+    private var isPrivate: Bool!
+    private var dismissing = false
+    var tabManager: TabManager!
```
In general try to avoid using optionals as much as you can unless you need them to convey an empty state. For example, since we pass in `profile` in the constructor, that doesn't need to be an optional. Additionally, try to avoid explicit unwrapped optionals. They tend to be a source of exceptions so I find minimizing their use helps a lot. Some solutions include using `lazy` properties or using regular optionals if you're not sure the value will exist.

sleroux commented on Client/Frontend/Browser/BackForwardListView.swift:
```diff
@@ -0,0 +1,196 @@
+        tableView.separatorStyle = .None
+        tableView.dataSource = self
+        tableView.delegate = self
+        tableView.tableFooterView = UIView(frame: CGRectMake(0, 0, tableView.frame.size.width, 1))
```
Nit: Is this the trick to remove the lines past the bottom? I think you can just use `UIView()` with no frame.

sleroux commented on Client/Frontend/Browser/BackForwardListView.swift:
```diff
@@ -0,0 +1,196 @@
+        tableView.alwaysBounceVertical = false
+        
+        if isPrivate {
+            tableView.backgroundColor = UIColor.init(colorLiteralRed: 0.5, green: 0.5, blue: 0.5, alpha: 0.4)
```
Nit: Don't need to call `init`, just `UIColor(blah)`

sleroux commented on Client/Frontend/Browser/BackForwardListView.swift:
```diff
@@ -0,0 +1,196 @@
+            tableView.backgroundColor = UIColor.init(colorLiteralRed: 1, green: 1, blue: 1, alpha: 0.4)
+            let blurEffect = UIBlurEffect(style: .ExtraLight)
+            let blurEffectView = UIVisualEffectView(effect: blurEffect)
+            tableView.backgroundView = blurEffectView
```
Nit: The code in both branches of this `if` statement are the same. Maybe refactor into a small helper functiona?

sleroux commented on Client/Frontend/Browser/BackForwardListView.swift:
```diff
@@ -0,0 +1,196 @@
+        
+        listData = listData.filter { !(($0.item.title ?? "").isEmpty)}
+        
+        let deferred = sql.getSitesForURLs(urls)
```
Nit: If you don't need a reference to the deferred you can just call `sql.getSitesForURLS(urls).uponQueue()`

sleroux commented on Client/Frontend/Browser/BackForwardListView.swift:
```diff
@@ -0,0 +1,196 @@
+            }
+            self.backgroundColor = UIColor.init(colorLiteralRed: 0, green: 0, blue: 0, alpha: 0.2)
+            self.layoutIfNeeded()
+        })
```
There shouldn't be any animation calls inside a `layoutSubviews` since this method can get call on every rotation and layout validation. We should look into alternative ways of animating the height of the `tableView`.

sleroux commented on Client/Frontend/Browser/BackForwardListView.swift:
```diff
@@ -0,0 +1,196 @@
+        }
+        
+        if currentRow > 1 {
+            self.tableView.reloadData()
```
This `reloadData` won't do much since the data is asynchronously fetched above. Don't think we need it?

sleroux commented on Client/Frontend/Browser/BackForwardListView.swift:
```diff
@@ -0,0 +1,196 @@
+            let moveToIndexPath = NSIndexPath(forRow: currentRow-2, inSection: 0)
+            self.tableView.reloadRowsAtIndexPaths([moveToIndexPath], withRowAnimation: .None)
+            self.tableView.scrollToRowAtIndexPath(moveToIndexPath, atScrollPosition: UITableViewScrollPosition.Middle, animated: false)
+        }
```
Nit: In general this init method does a lot of things. Might be easier to break it apart into small, single purpose functions to bring some clarity.

sleroux commented on Client/Frontend/Browser/BackForwardListView.swift:
```diff
@@ -0,0 +1,196 @@
+    }
+    
+    func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
+        var cell: BackForwardTableViewCell? = tableView.dequeueReusableCellWithIdentifier("BackForwardListViewController") as? BackForwardTableViewCell
```
This is the older way of dequeuing cells. If you use the method that takes in an `NSIndexPath`, you don't need to do the nil check.

sleroux replied:
Also safe to use `as` instead of `as?` for the dequeue call.

sleroux commented on Client/Frontend/Browser/BackForwardListView.swift:
```diff
@@ -0,0 +1,196 @@
+        cell?.connectingBackwards = (indexPath.item != listData.count-1)
+        cell?.connectingForwards = (indexPath.item != 0)
+        
+        cell?.setNeedsDisplay()
```
Shouldn't need this.

sleroux commented on Client/Frontend/Browser/BackForwardListView.swift:
```diff
@@ -0,0 +1,196 @@
+        return CGFloat(BackForwardViewUX.RowHeight);
+    }
+    
+    override func pointInside(point: CGPoint, withEvent event: UIEvent?) -> Bool {
```
Is this needed if the user taps outside of the table view? I thought the table view was the entire view? Even if it isn't it might be better to just use a `UITapGestureRecognizer` instead of having to check if a point is within a frame.

sleroux commented on Client/Frontend/Browser/BackForwardTableViewCell.swift:
```diff
@@ -0,0 +1,149 @@
+//
+//  Created by Tyler Lacroix on 5/17/16.
+//  Copyright © 2016 Mozilla. All rights reserved.
+//
```
Nit: Use MPL preamble.

sleroux commented on Client/Frontend/Browser/BackForwardTableViewCell.swift:
```diff
@@ -0,0 +1,149 @@
+    
+    var faviconView: UIImageView!
+    var label: UILabel!
+    var bg: UIView!
```
Nit: `!`

sleroux commented on Client/Frontend/Browser/BackForwardTableViewCell.swift:
```diff
@@ -0,0 +1,149 @@
+    var connectingForwards = true
+    var connectingBackwards = true
+    
+    var currentTab = false  {
```
I would rename this to `isCurrentTab` to help indicate it's a boolean since we use `currentTab` across the app to store the current tab instance.

sleroux commented on Client/Frontend/Browser/BackForwardTableViewCell.swift:
```diff
@@ -0,0 +1,149 @@
+        }
+        else {
+            label.textColor = UIColor.blackColor()
+        }
```
Nit: `label.textColor = isPrivate ? .whiteColor : .blackColor()

sleroux commented on Client/Frontend/Browser/BackForwardTableViewCell.swift:
```diff
@@ -0,0 +1,149 @@
+        bg.backgroundColor = BackForwardViewCellUX.bgColor
+        
+        self.addSubview(bg)
+        self.sendSubviewToBack(bg)
```
Instead of using `sendSubviewToBack`, it might be better to add the subviews in the order that you want them to appear.

sleroux commented on Client/Frontend/Browser/BackForwardTableViewCell.swift:
```diff
@@ -0,0 +1,149 @@
+    }
+    
+    override func drawRect(rect: CGRect) {
+        let context = UIGraphicsGetCurrentContext();
```
Whenever overriding framework methods like this, I always try to call the super method to make sure the framework does what it needs to do first.

sleroux commented on Storage/SQL/SQLiteHistory.swift:
```diff
@@ -673,6 +673,15 @@ extension SQLiteHistory: Favicons {
         return db.runQuery(sql, args: args, factory: SQLiteHistory.iconColumnFactory)
     }
+    
+    public func getSitesForURLs(urls: [String]) -> Deferred<Maybe<Cursor<Site?>>> {
```
👌

sleroux commented on Client/Frontend/Browser/BackForwardListViewController.swift:
```diff
@@ -4,56 +4,203 @@
+            }
+        }
+        
+        if currentRow > 1 {
```
Nit: Instead of having the `if` check here, it would be safer to include this check as a `guard` inside the `scrollTableViewToIndex` method in case we call the method from somewhere else and forget to do the check.

sleroux commented on Client/Frontend/Browser/BackForwardListViewController.swift:
```diff
@@ -4,56 +4,203 @@
+        self.tableView.scrollToRowAtIndexPath(moveToIndexPath, atScrollPosition: UITableViewScrollPosition.Middle, animated: false)
+    }
+    
+    override func didRotateFromInterfaceOrientation(fromInterfaceOrientation: UIInterfaceOrientation) {
```
This method is deprecated. You should use `viewWillTransitionToSize:` instead.

sleroux commented on Client/Frontend/Browser/BackForwardListViewController.swift:
```diff
@@ -4,56 +4,203 @@
+            make.height.equalTo(0)
             make.left.right.bottom.equalTo(self.view)
         }
+        self.view.layoutIfNeeded()
```
Nit: Don't need this layout if needed.

tylerlacroix replied:
I think this layoutIfNeeded is needed for the animation because it's creating the constraints and then animating the height in the same call.  Otherwise it seems to animate from the origin (0,0,0,0) to open, it seems to ignore the "snp_makeConstraints".

sleroux commented on Client/Frontend/Browser/BackForwardListViewController.swift:
```diff
@@ -4,56 +4,203 @@
+        let tap = UITapGestureRecognizer(target: self, action: #selector(BackForwardListViewController.handleTap))
+        tap.cancelsTouchesInView = false
+        tap.delegate = self
+        parentViewController?.view.addGestureRecognizer(tap)
```
How come you're adding the gesture to the parent view controller and not this view controller?

tylerlacroix replied:
The reason why I added the gesture to the parent, and why I added the view/view controller directly was to preserve the ability to use the menu UI with the long back list showing, for example if a user holds the back button, then taps the back button. I think presenting the view controller would lose this, but I can fix it if that's not the desired behavior :)

sleroux replied:
Visually does that mean that the menu would appear on top of the back forward list? I feel like the menu and back forward list would be mutually exclusive, ie., if we're showing the back forward list, dismiss, then show the menu. @tecgirl would we want the menu to overlap on top of the back forward list?

tylerlacroix replied:
Oh I see what you mean. From the mock-ups it looked like it slid from behind the menu, maybe I'm mistaken.

sleroux replied:
Ah I didn't know there were mock ups. If that's the case nevermind!

sleroux commented on Client/Frontend/Browser/BackForwardTableViewCell.swift:
```diff
@@ -0,0 +1,136 @@
+        contentView.addSubview(faviconView)
+        
+        label.text = " "
+        label.font = label.font.fontWithSize(BackForwardViewCellUX.fontSize)
```
Minor nit: Now that label/images are loaded, you can move the code that setups the properties on them inside a lazy initializer:

```swift
lazy var label: UILabel = {
    let label = UILabel()
    label.text = " "
    label.font = label.font.fontWithSize(BackForwardViewCellUX.fontSize)
    return label
}()
```

tylerlacroix replied:
Oh, smart!

sleroux commented on Client/Frontend/Browser/BrowserViewController.swift:
```diff
@@ -1642,6 +1634,20 @@ extension BrowserViewController: TabToolbarDelegate {
+                make.left.right.bottom.equalTo(self.webViewContainer)
+                make.left.right.bottom.equalTo(self.webViewContainer)
+            }
+        }
```
Instead of adding the view/view controller directly, it might be better to present the back forward list in a similar way we present the menu by _presenting_ the view controller instead of the view. We don't need the fancy animations the menu has but by implementing a custom `UIViewControllerAnimatedTransitioning` we can store the animation specific code for animating in/out the back forward list separately from the view controller. The `MenuPresentationAnimator` and `BrowserTrayAnimator` classes are more complex versions of this but you can use them as a reference for how we show/hide those components.

sleroux commented on Client/Frontend/Browser/BackForwardListAnimator.swift:
```diff
@@ -0,0 +1,80 @@
+    private func animateWithBackForward(backForward: BackForwardListViewController, browserViewController bvc: BrowserViewController, transitionContext: UIViewControllerContextTransitioning) {
+        let containerView = transitionContext.containerView();
+        
+        if(presenting == true) {
```
Remove brackets and just do `if presenting {`

sleroux commented on Client/Frontend/Browser/BackForwardListAnimator.swift:
```diff
@@ -0,0 +1,80 @@
+                }, completion: { (completed) -> Void in
+                    transitionContext.completeTransition(completed)
+            })
+            
```
Nit: Replace any uses of `!` with a `guard` at the top to prevent potential NPEs.

sleroux commented on Client/Frontend/Browser/BackForwardListViewController.swift:
```diff
@@ -84,64 +105,72 @@ class BackForwardListViewController: UIViewController, UITableViewDataSource, UI
+        dismissViewControllerAnimated(true, completion: nil)
     }
     
     func gestureRecognizer(gestureRecognizer: UIGestureRecognizer, shouldReceiveTouch touch: UITouch) -> Bool {
```
Not sure if you need to implement this method anymore.

tylerlacroix replied:
Yeah, this is needed, the OverCurrentContext style doesn't dismiss when touched outside automatically.

sleroux commented on Client/Frontend/Browser/BackForwardListViewController.swift:
```diff
@@ -24,12 +24,34 @@ class BackForwardListViewController: UIViewController, UITableViewDataSource, UI
+    var tableHeight: CGFloat
+    {
+        get {
+            return min(CGFloat(BackForwardViewUX.RowHeight*listData.count), self.view.frame.height/2)
```
We've run into problems in the pass with wrapping computed properties over UIKit APIs like this in the past when the caller was not on the main thread. Externally this looks like a basic `CGFloat` when it fact it reaches into UIKit which requires the main thread. I would add an `assert(NSThread.isMainThread(), ...)` check here to catch any places this might happen.

sleroux commented on Client/Frontend/Browser/BackForwardListViewController.swift:
```diff
@@ -24,12 +24,34 @@ class BackForwardListViewController: UIViewController, UITableViewDataSource, UI
+    }()
+    
     var tabManager: TabManager!
+    var bvc: BrowserViewController!
```
I'm wary of passing a reference to the BVC and retaining it in this `BackForwardListViewController`. From an ownership perspective, I can see the BVC's navigation controller to own a reference to this but I don't expect this to retain one back. I would make this a `weak` ref at the very least.

sleroux commented on Client/Frontend/Browser/BackForwardListViewController.swift:
```diff
@@ -84,64 +105,72 @@ class BackForwardListViewController: UIViewController, UITableViewDataSource, UI
+        tableView.snp_updateConstraints { make in
+            make.height.equalTo(min(CGFloat(BackForwardViewUX.RowHeight*listData.count), size.height/2))
+        }
+        if let superView = view.superview {
```
If we're setting this view's bounds to the edges either way, do we need to do this check every time we transition size? If we setup the constraints at the start we don't need to reset them.

sleroux commented on Client/Frontend/Browser/BackForwardListViewController.swift:
```diff
@@ -168,39 +197,18 @@ class BackForwardListViewController: UIViewController, UITableViewDataSource, UI
         cell.connectingForwards = (indexPath.item != 0)
         cell.isPrivate = isPrivate
         
+        cell.setNeedsDisplay()
```
I might have forgotten but why do we need this call again?

tylerlacroix replied:
This call is needed because the cells draw the connecting lines in the drawRect but sometimes if the cell is reused it won't redraw the correct line (whether it should connect up, down or both ways).

sleroux commented on Client/Frontend/Browser/BackForwardListViewController.swift:
```diff
@@ -24,12 +24,34 @@ class BackForwardListViewController: UIViewController, UITableViewDataSource, UI
+        }
+    }
+    
+    var backForwardTransitionDelegate: UIViewControllerTransitioningDelegate? {
```
We can probably remove this wrapper property and assign directly to `self.transitioningDelegate`. I don't see the advantage of giving this another name if that's all we're doing.

tylerlacroix replied:
This I also found strange, but it seems the transitioningDelegate is only a weak property, so the transitioningDelegate property is immediately set to nil if the object is not held. I saw this in MenuViewController so I thought I should keep the style of holding transitioning delegates the same.

sleroux commented:
Additional comments:

* Instead of calling `setupTableView` in the `init` call, I would move it to `viewDidLoad` to align it better with the view controller’s lifecycle with views. You could also move most of the logic inside the `setupTableView` call into a lazy load property on the VC. Just need to make sure not to call tableView prior to `viewDidLoad`.

* Looking at the menu in RTL pseudolanguage, it looks like the text isn't aligned correctly on the left side

![simulator screen shot may 31 2016 3 11 00 pm](https://cloud.githubusercontent.com/assets/434322/15687543/54518622-2743-11e6-8095-9f190617f811.png)

tylerlacroix commented:
Thanks Steph!  I added the changes and added RTL 
<img width="487" alt="screen shot 2016-05-31 at 4 43 41 pm" src="https://cloud.githubusercontent.com/assets/11801330/15689878/e9e71426-274e-11e6-970b-a1cceedee50b.png">

sleroux commented on Client/Frontend/Browser/BackForwardListViewController.swift:
```diff
@@ -36,12 +50,13 @@ class BackForwardListViewController: UIViewController, UITableViewDataSource, UI
     
     var tabManager: TabManager!
-    var bvc: BrowserViewController!
+    weak var bvc: BrowserViewController!
```
Since this is `weak` now, there is no guarentee that this will always have a value so we should  `!` ➡️ `?`

sleroux commented:
@tylerlacroix looks like this got out of sync with master. Could you rebase? I'll merge it in afterwards.

tylerlacroix commented:
Okay, it should be rebased :)
